src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+use super::Collapse;
+use smallvec::SmallVec;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// Include raw addresses (e.g., `0xbfff0836`) where symbols can't be found.
+    pub include_addrs: bool,
+
+    /// Annotate kernel functions with a `_[k]` suffix.
+    pub annotate_kernel: bool,
+}
+
+/// A stack collapser for the output of `perf script`.
+///
+/// To construct one, either use `perf::Folder::default()` or create an [`Options`] and use
+/// `perf::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
+
+    /// Skip all stack lines in this event.
+    skip_stack: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+    count: usize,
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Current comm name.
+    ///
+    /// Called pname after original stackcollapse-perf source.
+    pname: Option<String>,
+
+    //    event_filtering: EventFilterState,
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        //TODO: clean this up, we skip the first 4 lines:
```

Comment:
Looks like upstream defines this [as a constant](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L53) and then just skips that many lines. We should probably do something similar.

Reply:
See also the comment below about simply ignoring empty lines.

Reply:
I updated this to skip a given number of lines.

Reply:
I think the strategy [below](https://github.com/jonhoo/inferno/pull/78/commits/bdf75bfe0183a32ec8d6d2196e70b18a7d3512e3#r264005789) is probably a better solution.